In [25]:
#import sys
#!{sys.executable} -m pip install haversine
import haversine 
from sklearn.neighbors import DistanceMetric
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
plt.style.use("dark_background")
plt.rcParams['figure.dpi'] = 150
from nltk import flatten

We read the santander_locations document, which contains the location information of bike stations across London.

In [2]:
station_data = pd.read_csv("./santander_locations.csv")

In [3]:
station_data.head(10)

Station.Id                           StationName  longitude   latitude  \
0           1             River Street, Clerkenwell  -0.109971  51.529200   
1           2        Phillimore Gardens, Kensington  -0.197574  51.499600   
2           3  Christopher Street, Liverpool Street  -0.084606  51.521300   
3           4       St. Chad's Street, King's Cross  -0.120974  51.530100   
4           5         Sedding Street, Sloane Square  -0.156876  51.493100   
5           6        Broadcasting House, Marylebone  -0.144229  51.518100   
6           7     Charlbert Street, St. John's Wood  -0.168074  51.534300   
7           8                Maida Vale, Maida Vale  -0.183486  51.529857   
8           9              New Globe Walk, Bankside  -0.096441  51.507400   
9          10                 Park Street, Bankside  -0.092754  51.506000   

      Easting    Northing  
0  531202.520  182832.020  
1  525207.070  179391.860  
2  532984.810  182001.530  
3  530436.760  182911.990  
4  528051.649  178742.097  
5  528857.440  181542.870  
6  527158.010  183300.750  
7  526102.000  182780.000  
8  532203.970  180434.550  
9  532463.890  180284.300

In [61]:
len(station_data)

808

The number of stations is 808.

In [1]:
station_data['Station.Id'].to_numpy()

NameError: name 'station_data' is not defined

The leftmost column in bold gives the index of the table. It is not the same as Station ID, which is given by the column named 'Station.Id'. But we could always access entries by Station ID, or by the table ID.

In [59]:
#Access by Station ID
station_data[station_data['Station.Id'] == 268].index

Int64Index([261], dtype='int64')

In [60]:
#Access by Table ID
station_data.iloc[261]

Station.Id                         268
StationName    Belgrave Road, Victoria
longitude                    -0.144133
latitude                       51.4932
Easting                         528934
Northing                        178773
Name: 261, dtype: object

For the station Belgrave Road, Victoria, its station ID is 268, but its table ID is 261.

In [4]:
#Access by Station ID
station_data[station_data['StationName'] == "Boston Place, Marylebone"].index

Int64Index([43], dtype='int64')

In [6]:
#Access by Table ID
station_data.iloc[766]

Station.Id                                 798
StationName    Birkenhead Street, King's Cross
longitude                            -0.122299
latitude                               51.5302
Easting                                 530345
Northing                                182925
Name: 766, dtype: object

Below gives a collection of bike stations near National Railway stations.

| Station names | Station ID | DF Index | 
| --- | --- | --- |
| Belgrave Road, Victoria | 268 | 261 |
| Waterloo Station 1, Waterloo | 374 | 362 |
| Waterloo Station 2, Waterloo | 261 | 350 |
| Waterloo Station 3, Waterloo | 154 | 151 |
| South Wharf Road, Paddington | 186 | 182 |
| Birkenhead Street, King's Cross | 798 | 766 |
| Snowsfields, London Bridge | 706 | 675 |
| Boston Place, Marylebone | 45 | 43 |

We now do some processing of the data.

In [9]:
max(station_data['Station.Id']), len(np.unique(station_data['Station.Id']))
station_ID = list(station_data['Station.Id'])
row_no = list(range(len(station_data['Station.Id'])))
row_ID_dict = dict(zip(station_ID, row_no))

In [10]:
class OptimizationError(RuntimeError):
    """Called when optimizer does not converge."""
    pass

class StationIdError(IndexError):
    """Called when we try and read a non-existing station id."""
    pass

We write a function to obtain station name corresponding to a given station ID number.

In [11]:
def get_station_name(in_id):
    """Get station name from bike_data for a given id."""
    try:
        return station_data[station_data["Station.Id"] == in_id].StationName.iloc[0]
    except IndexError:
        StationIdError("No station matching input ID")

In [12]:
get_station_name(830)

'Sidney Road, Stockwell'

We get the longitude and latitude data in radians from the data.

In [13]:
from math import radians
station_coord = station_data[["longitude","latitude"]]
station_coord['longitude_radians'] = station_coord['longitude'].apply(lambda x: radians(x)) 
station_coord['latitude_radians'] = station_coord['latitude'].apply(lambda x: radians(x)) 
station_coord = station_coord[['latitude_radians','longitude_radians']].to_numpy()

/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until
/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  after removing the cwd from sys.path.


In [14]:
station_coord

array([[ 8.99354201e-01, -1.91935603e-03],
       [ 8.98837583e-01, -3.44831682e-03],
       [ 8.99216320e-01, -1.47664803e-03],
       ...,
       [ 8.99038384e-01, -1.89027885e-03],
       [ 8.98821928e-01, -8.56206171e-04],
       [ 8.98811351e-01, -1.10744632e-03]])

And we find the geodesic distance between each station, using the haversine formula.

In [ ]:
dist = DistanceMetric.get_metric('haversine')
geo_dist = dist.pairwise(station_coord) * 6365.079
geo_dist_df = pd.DataFrame(geo_dist)
geo_dist_df.iloc[1,2]

We save the geodesic distance between each pair of stations into an array.

In [16]:
geo_dist

array([[ 0.        ,  6.891488  ,  1.9605899 , ...,  2.01349835,
         5.4049247 ,  4.72041716],
       [ 6.891488  ,  0.        ,  8.17421589, ...,  6.30370511,
        10.271535  ,  9.27709531],
       [ 1.9605899 ,  8.17421589,  0.        , ...,  1.99173004,
         3.51324925,  2.96369879],
       ...,
       [ 2.01349835,  6.30370511,  1.99173004, ...,  0.        ,
         4.32238187,  3.42167331],
       [ 5.4049247 , 10.271535  ,  3.51324925, ...,  4.32238187,
         0.        ,  0.99781121],
       [ 4.72041716,  9.27709531,  2.96369879, ...,  3.42167331,
         0.99781121,  0.        ]])

We save the geodesic dataframe into csv.

In [18]:
geo_dist_df.to_csv('./geodesic.csv')

We read the bike data folder, create a dictionary, store the file name as key and the data frame inside the file as value.

In [19]:
path = './santander_summaries'
data_files = os.listdir(path)
n_weeks = len(data_files)
bike_data = {file: pd.read_csv(path + '/' + file,
                                     names=["start_loc", "end_loc",
                                            "start_t", "duration"])
             for file in data_files}        

print(list(bike_data.values())[0].head())

   start_loc  end_loc   start_t  duration
0         47       56  55728000      1500
1        762      600  55728000       540
2        588      698  55728060       420
3        456      456  55728120     12240
4         67       67  55728120       660


Then we sort the data according to start time.

In [3]:
bike_keys = sorted(bike_data)
sorted_bike_data = {i: bike_data[i] for i in bike_keys}
 
print(sorted_bike_data)

NameError: name 'bike_data' is not defined

We map the row number to each location index in station_data, and add the distance column for each journey.

In [50]:
bike_loc = dict(sorted_bike_data)

for names in bike_loc.keys():
    endt = np.zeros(bike_loc[names].shape[0], dtype = 'float64')
    bike_loc[names] = bike_loc[names].to_numpy()
    startt = bike_loc[names][:, 2].astype('float64')
    dura = bike_loc[names][:, 3].astype('float64')
    bike_loc[names] = bike_loc[names][:, 0:2]
        
    start_noise = np.random.uniform(0.0, 1.0, bike_loc[names].shape[0])
    startt = startt / 60 + start_noise
    duration_noise = np.random.uniform(0.0, 1.0, bike_loc[names].shape[0])
    dura = dura / 60 + duration_noise
    endt = startt + dura
    
    dist = np.zeros((bike_loc[names].shape[0], 1), dtype = geo_dist.dtype)
    for r in range(bike_loc[names].shape[0]):
            dist[r] = geo_dist[row_ID_dict[bike_loc[names][r, 0]], row_ID_dict[bike_loc[names][r, 1]]]
    
    dist = pd.DataFrame(dist)
    startt = pd.DataFrame(startt)
    endt = pd.DataFrame(endt)
    dura = pd.DataFrame(dura)
    bike_loc[names] = pd.DataFrame(bike_loc[names], columns = ["start_loc", "end_loc"])
    
    bike_loc[names] = bike_loc[names].assign(distance = dist)
    bike_loc[names] = bike_loc[names].assign(start_t = startt)
    bike_loc[names] = bike_loc[names].assign(end_t = endt)
    bike_loc[names] = bike_loc[names].assign(duration = dura)
    
print(list(bike_loc.values())[0].head(10))

   start_loc  end_loc  distance        start_t          end_t   duration
0        103       37  1.455347  787682.433041  787689.133564   6.700523
1         39      539  0.544400  787682.524014  787685.358116   2.834102
2        785      785  0.000000  787682.465639  787688.043628   5.577989
3        341      159  1.090538  787683.657119  787714.656619  30.999500
4        708      573  3.599904  787683.240691  787702.209320  18.968630
5        366      138  0.463223  787684.859478  787692.471952   7.612474
6         66      832  6.017458  787684.177849  787711.647774  27.469925
7        225      384  1.459768  787685.716766  787692.953482   7.236715
8        729      207  4.568847  787686.183963  787706.512693  20.328729
9        510      763  1.334991  787686.003770  787691.893567   5.889797


We repeat the above procedure, but normalise the data.

In [51]:
bike_loc = dict(sorted_bike_data)

for names in bike_loc.keys():
    bike_loc[names] = bike_loc[names].to_numpy()
    startt = bike_loc[names][:, 2].astype('float64')
    dura = bike_loc[names][:, 3].astype('float64')
    endt = startt + dura
    bike_loc[names] = bike_loc[names][:, 0:2]
    t_min = np.floor(47260920/(60*60*24))*60*60*24
        
    start_noise = np.random.uniform(0.0, 1.0, bike_loc[names].shape[0])
    startt = (startt - t_min) / 60 + start_noise
    end_noise = np.random.uniform(0.0, 1.0, bike_loc[names].shape[0])
    endt = (endt - t_min) / 60 + end_noise
    dura = endt - startt
    
    dist = np.zeros((bike_loc[names].shape[0], 1), dtype = geo_dist.dtype)
    for r in range(bike_loc[names].shape[0]):
            dist[r] = geo_dist[row_ID_dict[bike_loc[names][r, 0]], row_ID_dict[bike_loc[names][r, 1]]]
    
    dist = pd.DataFrame(dist)
    startt = pd.DataFrame(startt)
    endt = pd.DataFrame(endt)
    dura = pd.DataFrame(dura)
    bike_loc[names] = pd.DataFrame(bike_loc[names], columns = ["start_loc", "end_loc"])
    
    bike_loc[names] = bike_loc[names].assign(distance = dist)
    bike_loc[names] = bike_loc[names].assign(start_t = startt)
    bike_loc[names] = bike_loc[names].assign(end_t = endt)
    bike_loc[names] = bike_loc[names].assign(duration = dura)
    
print(list(bike_loc.values())[0].head(10))

   start_loc  end_loc  distance   start_t      end_t   duration
0        103       37  1.455347  2.119259   8.670248   6.550988
1         39      539  0.544400  2.140374   4.792660   2.652286
2        785      785  0.000000  2.088577   7.816521   5.727944
3        341      159  1.090538  3.504080  33.847014  30.342934
4        708      573  3.599904  3.048540  21.242909  18.194369
5        366      138  0.463223  4.027290  11.793456   7.766166
6         66      832  6.017458  4.414472  31.792052  27.377580
7        225      384  1.459768  5.242881  12.314880   7.072000
8        729      207  4.568847  6.529955  26.907634  20.377679
9        510      763  1.334991  6.181161  11.148302   4.967141


We obtain a dictionary containing all bike journeys that start from station ID 2, i.e., 'start_loc' == 2. And we also obtain a dictionary containing all bike journeys that end at station ID 2, i.e., 'end_loc' == 2.

In [52]:
data_dict_dep = {}
data_dict_arr = {}

for names in bike_loc.keys():
    data_dict_dep[names] = bike_loc[names].loc[bike_loc[names]['start_loc'] == 2]
    data_dict_arr[names] = bike_loc[names].loc[bike_loc[names]['end_loc'] == 2]
        
print(list(data_dict_dep.values())[0].head(10));
print(list(data_dict_arr.values())[0].head(10))

      start_loc  end_loc  distance     start_t       end_t    duration
423           2        2  0.000000  138.390326  167.524228   29.133902
424           2        2  0.000000  138.118309  164.339792   26.221483
428           2        2  0.000000  139.168505  166.533031   27.364525
3910          2      558  4.882449  511.574332  550.903712   39.329380
3983          2      228  4.438654  513.149723  530.649633   17.499910
4071          2      219  1.156326  516.339959  536.385236   20.045277
4089          2      219  1.156326  516.786659  536.875989   20.089330
4189          2      389  3.117940  519.214417  533.429633   14.215215
4454          2      826  3.789855  528.408782  948.495018  420.086236
5182          2      348  3.544449  553.272058  573.238550   19.966493
      start_loc  end_loc  distance     start_t       end_t   duration
423           2        2  0.000000  138.390326  167.524228  29.133902
424           2        2  0.000000  138.118309  164.339792  26.221483
428      

We save the departure times data and the arrival times data into a new folder.

In [53]:
os.makedirs('./SortDepartures', exist_ok=True)  

for names in data_dict_dep.keys():
    data_dict_dep[names].to_csv('./Departures/'+ str(names))

os.makedirs('./SortArrivals', exist_ok=True)  

for names in data_dict_arr.keys():
    data_dict_arr[names].to_csv('./Arrivals/'+ str(names))

We obtain a dictionary containing all bike journeys that start from station ID 2, 3, 4, 5, i.e., 'start_loc' == 2,3, 4, 5. 

In [54]:
loc_codes = [2, 3, 4, 5]
data_dict_0 = {}
data_dict_1 = {}
data_dict_2 = {}
data_dict_3 = {}

for names in bike_loc.keys():
    data_dict_0[names] = bike_loc[names].loc[bike_loc[names]['start_loc'] == 2]
    data_dict_1[names] = bike_loc[names].loc[bike_loc[names]['start_loc'] == 3]
    data_dict_2[names] = bike_loc[names].loc[bike_loc[names]['start_loc'] == 4]
    data_dict_3[names] = bike_loc[names].loc[bike_loc[names]['start_loc'] == 5]
        
print(list(data_dict_0.values())[0].head(10));
print(list(data_dict_1.values())[0].head(10))
print(list(data_dict_2.values())[0].head(10))
print(list(data_dict_3.values())[0].head(10))

      start_loc  end_loc  distance     start_t       end_t    duration
423           2        2  0.000000  138.390326  167.524228   29.133902
424           2        2  0.000000  138.118309  164.339792   26.221483
428           2        2  0.000000  139.168505  166.533031   27.364525
3910          2      558  4.882449  511.574332  550.903712   39.329380
3983          2      228  4.438654  513.149723  530.649633   17.499910
4071          2      219  1.156326  516.339959  536.385236   20.045277
4089          2      219  1.156326  516.786659  536.875989   20.089330
4189          2      389  3.117940  519.214417  533.429633   14.215215
4454          2      826  3.789855  528.408782  948.495018  420.086236
5182          2      348  3.544449  553.272058  573.238550   19.966493
       start_loc  end_loc  distance      start_t        end_t   duration
3999           3       57  2.470578   514.570754   527.917456  13.346702
7170           3        3  0.000000   641.234114   658.241234  17.007120


We save the departure times data at each station into a new folder.

In [237]:
os.makedirs('./Station2', exist_ok=True)  

for names in data_dict_dep.keys():
    data_dict_0[names].to_csv('./Station2/'+ str(names))

os.makedirs('./Station3', exist_ok=True)  

for names in data_dict_arr.keys():
    data_dict_1[names].to_csv('./Station3/'+ str(names))

os.makedirs('./Station4', exist_ok=True)  

for names in data_dict_arr.keys():
    data_dict_2[names].to_csv('./Station4/'+ str(names))
                                
os.makedirs('./Station5', exist_ok=True)  

for names in data_dict_arr.keys():
    data_dict_3[names].to_csv('./Station5/'+ str(names))

We then read and save departure times at all stations.

In [55]:
loc_codes = station_data['Station.Id'].to_numpy()

for i in range(808):
    
    loc = loc_codes[i]
    data_dict = {}

    for names in bike_loc.keys():
        data_dict[names] = bike_loc[names].loc[bike_loc[names]['start_loc'] == loc]
    
    path = f"./sortdep/Station{loc}"
    os.makedirs(path, exist_ok=True) 
    
    filename = f"./sortdep/Station{loc}/{str(names)}"
    for names in data_dict_dep.keys():
        filename = f"./sortdep/Station{loc}/{str(names)}"
        data_dict[names].to_csv(filename)

#Add in duration.
cols = ["start_t", "end_t","duration","start_loc", "end_loc"]
for name in bike_data.keys():
    bike_data[name] = bike_data[name].assign(end_t=lambda x:
                                             abs(x.start_t + x.duration))
    bike_data[name] = bike_data[name].reindex(columns=cols)

print(list(bike_data.values())[0].head())

#Map the row number to each location index in station_data.
for name in bike_data.keys():
    bike_data[name] = bike_data[name].to_numpy()
    startno = np.zeros((bike_data[name].shape[0], 1), dtype=int)
    endno = np.zeros((bike_data[name].shape[0], 1), dtype=int)
    
    for r in range(bike_data[name].shape[0]):
        startno[r] = row_ID_dict[bike_data[name][r, 3]]
        endno[r] = row_ID_dict[bike_data[name][r, 4]]
    
    startno = pd.DataFrame(startno)
    endno = pd.DataFrame(endno)
    bike_data[name] = pd.DataFrame(bike_data[name], columns = ["start_t", "end_t","duration","start_loc", 
                                                              "end_loc"])
    
    bike_data[name] = bike_data[name].assign(start_no = startno)
    bike_data[name] = bike_data[name].assign(end_no = endno)
    
print(list(bike_data.values())[0].head())        

And then we add the distance travelled for each journey.

In [105]:
#Adding the distance column for each journey
bike_loc = dict(bike_data)

for names in bike_loc.keys():
    bike_loc[names] = bike_loc[names].to_numpy()
     
    
    dist = np.zeros((bike_loc[names].shape[0], 1), dtype = geo_dist.dtype)
    for r in range(bike_loc[names].shape[0]):
            dist[r] = geo_dist[bike_loc[names][r, 5], bike_loc[names][r, 6]]
    
    dist = pd.DataFrame(dist)
    bike_loc[names] = pd.DataFrame(bike_loc[names], columns = ["start_t", "end_t","duration","start_loc", 
                                                              "end_loc", "start_no", "end_no"])
    del bike_loc[names]["start_no"]
    del bike_loc[names]["end_no"]
    bike_loc[names] = bike_loc[names].assign(distance = dist)
    
print(list(bike_loc.values())[0].head(10))

    start_t     end_t  duration  start_loc  end_loc  distance
0  55728000  55729500      1500         47       56  0.029266
1  55728000  55728540       540        762      600  0.019513
2  55728060  55728480       420        588      698  0.013609
3  55728120  55740360     12240        456      456  0.000000
4  55728120  55728780       660         67       67  0.000000
5  55728120  55728420       300        243       74  0.015976
6  55728180  55728480       300        715      444  0.011863
7  55728180  55740420     12240        456      456  0.000000
8  55728180  55740420     12240        456      456  0.000000
9  55728180  55728600       420        383       83  0.003483


Finally we save the data into folders.

In [56]:
loc_codes = station_data['Station.Id'].to_numpy()

for i in range(808):
    
    loc = loc_codes[i]
    data_dict = {}

    for names in bike_loc.keys():
        data_dict[names] = bike_loc[names].loc[bike_loc[names]['start_loc'] == loc]
    
    path = f"./sortarr/Station{loc}"
    os.makedirs(path, exist_ok=True) 
    
    filename = f"./sortarr/Station{loc}/{str(names)}"
    for names in data_dict_dep.keys():
        filename = f"./sortarr/Station{loc}/{str(names)}"
        data_dict[names].to_csv(filename)